src/Control/Monad/Logic/Sequence/Internal.hs

```diff
@@ -198,30 +258,90 @@ instance Monad m => Applicative (SeqT m) where
 instance Monad m => Alternative (SeqT m) where
   {-# INLINE empty #-}
   {-# INLINEABLE (<|>) #-}
-  {-# SPECIALIZE INLINE (<|>) :: Seq a -> Seq a -> Seq a #-}
   empty = SeqT S.empty
-  m <|> n = fromView (altView m n)
+  (<|>) = alt
 
-altView :: Monad m => SeqT m a -> SeqT m a -> m (View m a)
-altView (toView -> m) n = m >>= \x -> case x of
-  Empty -> toView n
-  h :< t -> return (h :< cat t n)
-    where cat (SeqT l) (SeqT r) = SeqT (l S.>< r)
-{-# INLINE altView #-}
+alt :: Monad m => SeqT m a -> SeqT m a -> SeqT m a
+alt a b = unstream (alt_s (stream a) (stream b))
+{-# INLINE[3] alt #-}
+
+alt_s :: Monad m => StreamM m a -> StreamM m a -> StreamM m a
+alt_s (StreamM next_a a) (StreamM next_b b) = StreamM next (Just a, b)
+  where
+  {-# INLINE next #-}
+  next (Nothing, s_b) = do
+    x <- next_b s_b
+    return $ case x of
+      Done -> Done
+      Skip vs -> Skip (Nothing, vs)
+      Yield v vs -> Yield v (Nothing, vs)
+  next (Just s_a, s_b) = do
+    y <- next_a s_a
+    case y of
+      Done -> do
+        x <- next_b s_b
+        return $ case x of
+          Done -> Done
+          Skip vs -> Skip (Nothing, vs)
+          Yield v vs -> Yield v (Nothing, vs)
+      Skip vs -> return (Skip (Just vs, s_b))
+      Yield x xs -> return (Yield x (Just xs, s_b))
+{-# INLINE[1] alt_s #-}
 
 instance Monad m => Monad (SeqT m) where
   {-# INLINE return #-}
   {-# INLINEABLE (>>=) #-}
-  {-# SPECIALIZE INLINE (>>=) :: Seq a -> (a -> Seq b) -> Seq b #-}
   return = fromView . single
-  (toView -> m) >>= f = fromView $ m >>= \x -> case x of
-    Empty -> return Empty
-    h :< t -> f h `altView` (t >>= f)
-
-  {-# INLINEABLE (>>) #-}
-  (toView -> m) >> n = fromView $ m >>= \x -> case x of
-    Empty -> return Empty
-    _ :< t -> n `altView` (t >> n)
+  (>>=) = bind
+
+bind :: Monad m => SeqT m a -> (a -> SeqT m b) -> SeqT m b
+bind m f = unstream (bind_s (stream m) (stream . f))
+{-# INLINE[3] bind #-}
+
+done :: Monad m => StreamM m a
+done = StreamM (const (return Done)) Empty
+{-# INLINE CONLIKE [0] done #-}
```

Comment:
I think `CONLIKE` is justifiable here given that everything here is non-polymorphic and concretely specified. So no surprises right?